apps/rush-lib/src/cli/actions/CustomRushAction.ts

```diff
@@ -0,0 +1,175 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import * as fsx from 'fs-extra';
+import * as os from 'os';
+import * as colors from 'colors';
+
+import {
+  ICustomOption,
+  ICustomEnumOption,
+  ICustomEnumValue,
+  Stopwatch,
+  Event
+} from '../../index';
+
+import {
+  CommandLineFlagParameter,
+  CommandLineIntegerParameter,
+  CommandLineStringListParameter,
+  CommandLineOptionParameter,
+  ICommandLineActionOptions
+} from '@microsoft/ts-command-line';
+
+import RushCommandLineParser from './RushCommandLineParser';
+import { BaseRushAction } from './BaseRushAction';
+import { TaskManager } from '../utilities/TaskManager';
+
+interface ICustomOptionInstance extends ICustomOption {
+  longName: string;
+  parameterValue: CommandLineFlagParameter | CommandLineOptionParameter;
+}
+
+export class CustomRushAction extends BaseRushAction {
+  private customOptions: Array<ICustomOptionInstance> = [];
+
+  private _fromFlag: CommandLineStringListParameter;
+  private _toFlag: CommandLineStringListParameter;
+  private _verboseParameter: CommandLineFlagParameter;
+  private _parallelismParameter: CommandLineIntegerParameter;
+
+  private _parser: RushCommandLineParser;
+
+  constructor(parser: RushCommandLineParser, options: ICommandLineActionOptions) {
+    super(options);
+    this._parser = parser;
+  }
+
+  public addCustomOption(longName: string, option: ICustomOption): void {
+    const optionInstance: ICustomOptionInstance = option as ICustomOptionInstance;
+    optionInstance.longName = longName;
+    this.customOptions.push(optionInstance);
+  }
+
+  public run(): void {
+    if (!fsx.existsSync(this.rushConfiguration.rushLinkJsonFilename)) {
+      throw new Error(`File not found: ${this.rushConfiguration.rushLinkJsonFilename}` +
+        `${os.EOL}Did you run "rush link"?`);
+    }
+    this.eventHooksManager.handle(Event.preRushBuild);
+
+    const stopwatch: Stopwatch = Stopwatch.start();
+
+    const isQuietMode: boolean = !(this._verboseParameter.value);
+
+    // collect all custom flags here
+    const customFlags: string[] = [];
+    this.customOptions.forEach((customOption: ICustomOptionInstance) => {
+      if (customOption.parameterValue.value) {
+        if (customOption.optionType === 'flag') {
+          customFlags.push(customOption.longName);
+        } else if (customOption.optionType === 'enum') {
+          customFlags.push(`${customOption.longName}=${customOption.parameterValue.value}`);
+        }
+      }
+    });
+
+    const tasks: TaskManager = new TaskManager(
+      this._parser.rushConfig,
+      this._toFlag.value,
+      this._fromFlag.value,
+      this.options.actionVerb,
+      customFlags,
+      isQuietMode,
+      this._parallelismParameter.value,
+      this.options.actionVerb === 'build'
```

Comment:
If this can't be simplified, maybe at least put it in an options interface.  Otherwise it will be easy to accidentally mix up the parameters